From 6968dcd5131a22f96dddc1fb55bd5902b16d9db1 Mon Sep 17 00:00:00 2001
From: Chunming Zhou <David1.Zhou@amd.com>
Date: Fri, 10 Mar 2017 14:48:09 +0800
Subject: [PATCH 1501/2094] drm/amd/powerplay/rv: power up/down sdma via the
 SMU

sdma is powered down by default in vbios,
need to power up in driver init.  Power it down
again on driver tear down.

Signed-off-by: Chunming Zhou <David1.Zhou@amd.com>
Reviewed-by: Hawking Zhang <Hawking.Zhang@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
---
 drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c b/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
index 6b6b755..6c931ba 100644
--- a/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
+++ b/drivers/gpu/drm/amd/powerplay/smumgr/rv_smumgr.c
@@ -243,6 +243,26 @@ static int rv_verify_smc_interface(struct pp_smumgr *smumgr)
 	return 0;
 }
 
+/* sdma is disabled by default in vbios, need to re-enable in driver */
+static int rv_smc_enable_sdma(struct pp_smumgr *smumgr)
+{
+	PP_ASSERT_WITH_CODE(!rv_send_msg_to_smc(smumgr,
+			PPSMC_MSG_PowerUpSdma),
+			"Attempt to power up sdma Failed!",
+			return -EINVAL);
+
+	return 0;
+}
+
+static int rv_smc_disable_sdma(struct pp_smumgr *smumgr)
+{
+	PP_ASSERT_WITH_CODE(!rv_send_msg_to_smc(smumgr,
+			PPSMC_MSG_PowerDownSdma),
+			"Attempt to power down sdma Failed!",
+			return -EINVAL);
+
+	return 0;
+}
 
 static int rv_smu_fini(struct pp_smumgr *smumgr)
 {
@@ -250,6 +270,7 @@ static int rv_smu_fini(struct pp_smumgr *smumgr)
 			(struct rv_smumgr *)(smumgr->backend);
 
 	if (priv) {
+		rv_smc_disable_sdma(smumgr);
 		cgs_free_gpu_mem(smumgr->device,
 				priv->smu_tables.entry[WMTABLE].handle);
 		cgs_free_gpu_mem(smumgr->device,
@@ -265,6 +286,8 @@ static int rv_start_smu(struct pp_smumgr *smumgr)
 {
 	if (rv_verify_smc_interface(smumgr))
 		return -EINVAL;
+	if (rv_smc_enable_sdma(smumgr))
+		return -EINVAL;
 
 	return 0;
 }
-- 
2.7.4

